src/plugins/unified_histogram/public/container/services/state_service.ts

```diff
@@ -0,0 +1,180 @@
+  /**
+   * The current query
+   */
+  query: Query | AggregateQuery;
+  /**
+   * The current request adapter used for non-Lens requests
+   */
+  requestAdapter: RequestAdapter | undefined;
+  /**
+   * The current search session ID
+   */
+  searchSessionId: string | undefined;
+  /**
+   * The current time interval of the chart
+   */
+  timeInterval: string;
+  /**
+   * The current time range
+   */
+  timeRange: TimeRange;
+  /**
+   * The current top panel height
+   */
+  topPanelHeight: number | undefined;
+  /**
+   * The current fetch status of the hits count request
+   */
+  totalHitsStatus: UnifiedHistogramFetchStatus;
+  /**
+   * The current result of the hits count request
+   */
+  totalHitsResult: number | Error | undefined;
+}
+
+/**
+ * The options used to initialize the comntainer state
+ */
+export interface UnifiedHistogramStateOptions {
+  /**
+   * The services required by the Unified Histogram components
+   */
+  services: UnifiedHistogramServices;
+  /**
+   * The prefix for the keys used in local storage -- leave undefined to avoid using local storage
+   */
+  localStorageKeyPrefix?: string;
+  /**
+   * The initial state of the container
+   */
+  initialState: Partial<UnifiedHistogramState> & Pick<UnifiedHistogramState, 'dataView'>;
+}
+
+export class UnifiedHistogramStateService {
+  private localStorageKeyPrefix?: string;
+  private services: UnifiedHistogramServices;
+  private state: UnifiedHistogramState;
+  private state$: BehaviorSubject<UnifiedHistogramState>;
+
+  constructor(options: UnifiedHistogramStateOptions) {
+    const { services, localStorageKeyPrefix, initialState } = options;
+
+    let chartHidden = false;
+    let topPanelHeight: number | undefined;
+    let breakdownField: string | undefined;
+
+    if (localStorageKeyPrefix) {
+      this.localStorageKeyPrefix = localStorageKeyPrefix;
+      chartHidden = getChartHidden(services.storage, localStorageKeyPrefix) ?? false;
+      topPanelHeight = getTopPanelHeight(services.storage, localStorageKeyPrefix);
+      breakdownField = getBreakdownField(services.storage, localStorageKeyPrefix);
+    }
+
+    this.services = services;
+    this.state = {
+      breakdownField,
+      chartHidden,
+      filters: [],
+      lensRequestAdapter: undefined,
+      query: services.data.query.queryString.getDefaultQuery(),
+      requestAdapter: undefined,
+      searchSessionId: undefined,
+      timeInterval: 'auto',
+      timeRange: services.data.query.timefilter.timefilter.getTimeDefaults(),
+      topPanelHeight,
+      totalHitsResult: undefined,
+      totalHitsStatus: UnifiedHistogramFetchStatus.uninitialized,
+      ...initialState,
+    };
+    this.state$ = new BehaviorSubject(this.state);
+  }
+
+  public getState$<T = UnifiedHistogramState>(
+    selector?: (state: UnifiedHistogramState) => T
+  ): Observable<T> {
+    if (selector) {
+      return this.state$.pipe(map(selector), distinctUntilChanged(isEqual));
```

Comment:
I got rid of the `getState$` function and instead just exposed the observable directly as `state$`, so no more use of `isEqual`. And I switched from a class-based state service to a `createStateService` function that returns a plain object so I could easily spread it over the API object to extend it.

Also, I added a `useStateSelector` function and updated the container code to use selectors instead of a subscription that stores the entire state in a state variable. It's not an exact duplicate of the one your linked to since I needed one that supports `undefined`, but it functions similarly.